File: streaming/src/main/java/com/example/streaming/service/FFmpegService.java
package com.example.streaming.service;

import net.bramp.ffmpeg.FFmpeg;
import net.bramp.ffmpeg.FFprobe;
import net.bramp.ffmpeg.builder.FFmpegBuilder;
import net.bramp.ffmpeg.probe.FFmpegProbeResult;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

@Service
public class FFmpegService {

    @Value("${ffmpeg.path}")
    private String ffmpegPath;

    @Value("${ffprobe.path}")
    private String ffprobePath;

    // convert video to HLS format
    public File convertToHls(File inputFile, String videoId) throws IOException {
        // create temp output directory
        Path outputDir = Files.createTempDirectory("hls_" + videoId);

        FFmpeg ffmpeg = new FFmpeg(ffmpegPath);

        FFmpegBuilder builder = new FFmpegBuilder()
                .setInput(inputFile.getAbsolutePath())
                .addOutput(outputDir.toString() + "/index.m3u8")
                .setFormat("hls")
                .addExtraArgs("-hls_time", "10") // 10 second chunks
                .addExtraArgs("-hls_list_size", "0") // keep all chunks
                .addExtraArgs("-hls_segment_filename",
                        outputDir.toString() + "/segment_%03d.ts")
                .addExtraArgs("-c:v", "libx264") // video codec
                .addExtraArgs("-c:a", "aac") // audio codec
                .addExtraArgs("-preset", "fast") // encoding speed
                .done();

        net.bramp.ffmpeg.FFmpegExecutor executor = new net.bramp.ffmpeg.FFmpegExecutor(ffmpeg);
        executor.createJob(builder).run();

        return outputDir.toFile();
    }

    // get video duration in seconds
    public int getDuration(File videoFile) throws IOException {
        FFprobe ffprobe = new FFprobe(ffprobePath);
        FFmpegProbeResult result = ffprobe.probe(videoFile.getAbsolutePath());
        return (int) result.getFormat().duration;
    }

    // generate thumbnail from video
    public File generateThumbnail(File videoFile, String videoId) throws IOException {
        Path outputDir = Files.createTempDirectory("thumb_" + videoId);
        String thumbnailPath = outputDir.toString() + "/thumbnail.jpg";

        FFmpeg ffmpeg = new FFmpeg(ffmpegPath);

        FFmpegBuilder builder = new FFmpegBuilder()
                .setInput(videoFile.getAbsolutePath())
                .addOutput(thumbnailPath)
                .setFrames(1) // extract 1 frame
                .addExtraArgs("-ss", "00:00:02") // at 2 seconds
                .addExtraArgs("-vf", "scale=640:360") // 640x360 size
                .done();

        net.bramp.ffmpeg.FFmpegExecutor executor = new net.bramp.ffmpeg.FFmpegExecutor(ffmpeg);
        executor.createJob(builder).run();

        return new File(thumbnailPath);
    }
}

File: streamnest-api/src/ffmpeg/ffmpeg.service.ts
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { execSync } from 'child_process';
import { mkdirSync, existsSync } from 'fs';
import { join } from 'path';
import * as os from 'os';

@Injectable()
export class FfmpegService {
  private ffmpegPath: string;
  private ffprobePath: string;

  constructor(private configService: ConfigService) {
    this.ffmpegPath = this.configService.get<string>('FFMPEG_PATH')!;
    this.ffprobePath = this.configService.get<string>('FFPROBE_PATH')!;
  }

  getDuration(inputPath: string): number {
    const result = execSync(
      `"${this.ffprobePath}" -v error -show_entries format=duration -of default=noprint_wrappers=1:nokey=1 "${inputPath}"`
    ).toString().trim();
    return Math.floor(parseFloat(result));
  }

  convertToHls(inputPath: string, videoId: number): string {
    const outputDir = join(os.tmpdir(), `hls_${videoId}`);
    if (!existsSync(outputDir)) mkdirSync(outputDir);

    execSync(
      `"${this.ffmpegPath}" -i "${inputPath}" -hls_time 10 -hls_list_size 0 ` +
      `-hls_segment_filename "${outputDir}/segment_%03d.ts" ` +
      `-c:v libx264 -c:a aac -preset fast "${outputDir}/index.m3u8"`
    );

    return outputDir;
  }

  generateThumbnail(inputPath: string, videoId: number): string {
    const outputDir = join(os.tmpdir(), `thumb_${videoId}`);
    if (!existsSync(outputDir)) mkdirSync(outputDir);

    const thumbnailPath = join(outputDir, 'thumbnail.jpg');

    execSync(
      `"${this.ffmpegPath}" -i "${inputPath}" -ss 00:00:02 -vframes 1 ` +
      `-vf scale=640:360 "${thumbnailPath}"`
    );

    return thumbnailPath;
  }
}